File: backend/app/services/sentry_scrub.py
"""Sentry privacy boundary: never export URL query strings or fragments."""

from __future__ import annotations

from typing import Any


def _strip(value: str) -> str:
    return value.split("?", 1)[0].split("#", 1)[0]


def before_send(event: dict[str, Any], _hint: dict[str, Any]) -> dict[str, Any]:
    request = event.get("request")
    if isinstance(request, dict):
        if isinstance(request.get("url"), str):
            request["url"] = _strip(request["url"])
        request.pop("query_string", None)
        headers = request.get("headers")
        if isinstance(headers, dict):
            for key, value in list(headers.items()):
                if str(key).lower() == "referer" and isinstance(value, str):
                    headers[key] = _strip(value)
    for crumb in event.get("breadcrumbs", {}).get("values", []):
        data = crumb.get("data") if isinstance(crumb, dict) else None
        if isinstance(data, dict):
            for key in ("url", "from", "to"):
                if isinstance(data.get(key), str):
                    data[key] = _strip(data[key])
    return event

File: frontend/src/lib/sentry-scrub.ts
/** Remove query strings and fragments before an error leaves the app. */

import type { ErrorEvent } from "@sentry/nextjs";

export function stripUrlSecrets(value: string): string {
  return value.split(/[?#]/, 1)[0] ?? value;
}

export function scrubSentryEvent(event: ErrorEvent): ErrorEvent {
  if (event.request) {
    if (typeof event.request.url === "string") {
      event.request.url = stripUrlSecrets(event.request.url);
    }
    delete event.request.query_string;
    for (const [key, value] of Object.entries(event.request.headers ?? {})) {
      if (key.toLowerCase() === "referer") {
        event.request.headers![key] = stripUrlSecrets(value);
      }
    }
  }
  for (const crumb of event.breadcrumbs ?? []) {
    for (const key of ["url", "from", "to"]) {
      const value = crumb.data?.[key];
      if (typeof value === "string") crumb.data![key] = stripUrlSecrets(value);
    }
  }
  return event;
}
